### PR TITLE
[NEW] Menu for changing push notification settings

### LINKED ISSUE #288: [NEW] Menu for changing push notification settings
Menu for changing the push notification on a per channel basis is missing. I think the app gets the settings from the configured settings on the web?
image

### PR DESCRIPTION
**Pull request by artrmz**
@RocketChat/ios

Closes #288 

- [x] add a controller for notifications preferences
- [x] switch cell
- [x] drop-down menu for a "choose cells"
- [x] implement the saveNotifications POST
- [x] localize everything
- [x] clean up the code
- [x] write tests
- [x] fix switch cell height when the description has more than 1 line
- [x] replace drop down menu with picker view (like system's add event)
- [x] notification's icon for subscription actions' list

- [OPTIONAL] allow changing global notification preferences (accessible from Preferences menu?)

NOTES:

- decide what to do with the getOneSubscription endpoint (I guess we have to implement it anyways to always get fresh settings from the server) // do we have to call it since Subscription is already being updated periodically in the app
- update subscription object after saving settings so Subscription object that is assigned to Chat/Messages is always up to date [?] // NOTE: seems to be already updated periodically

notifications

### CODE REVIEW

**Comment by rafaelks** [comment]
@artrmz Thank you for the PR! Can you take a look on the conflicts please?

**Comment by artrmz** [comment]
Resolved conflicts, although there is still lots of work left, sadly I did not have much time lately :( Hopefully, I'll get back to it soon.

**Comment by rafaelks** [comment]
@artrmz Thank you!

**Comment by Sameesunkaria** [comment]
Hey @artrmz !
I'm really looking forward to this feature. It would go well with #1504. 🎉

**Comment by artrmz** [comment]
@Sameesunkaria Hey man! I am still quite short on time but will do my best to push it forward a little bit this/next week.

**Comment by Sameesunkaria** [comment]
@artrmz I'm not very familiar with the API stuff rn. But if you need any help, I have plenty of time. 😃

**Comment by artrmz** [comment]
@Sameesunkaria Thank you! Pretty much API (getting current notifications settings and saving new ones) is the last thing to do ;) Seems likehttps://github.com/RocketChat/Rocket.Chat/pull/10128 has been merged so need to implement `channels.notifications` endpoint. I might implement it tonight if my kid lets me do that :D

**Comment by artrmz** [comment]
Tried to implement https://rocket.chat/docs/developer-guides/rest-api/subscriptions/getone/ but seems like values for notifications are missing or am I doing something wrong? @rafaelks

Getting such response (on open.rocket.chat):

```
Printing description of result.raw:
  "subscription" : {
    "name" : "ios-beta",
    "alert" : false,
    "_updatedAt" : "2018-04-14T22:30:40.275Z",
    "meta" : {
      "revision" : 48,
      "updated" : 1523745040279,
      "version" : 0,
      "created" : 1523628775399
    },
    "userMentions" : 0,
    "open" : true,
    "u" : {
      "name" : "Artrmz",
      "username" : "artrmz",
      "_id" : "GAcSEAjHQE43MDfud"
    },
    "groupMentions" : 0,
    "t" : "c",
    "ts" : "2018-04-06T16:40:54.884Z",
    "_id" : "evFXhSFJ8qnzizBo8",
    "rid" : "mD78xjvSctosNbSPr",
    "unread" : 0,
    "ls" : "2018-04-14T22:30:40.275Z",
    "customFields" : {

    },
    "fname" : "ios-beta"
  },
  "success" : true
}
```

**Comment by Sameesunkaria** [comment]
@artrmz I believe that only the preferences that are not set to default are sent.

This is what I get for any room.
```
{
  "subscription": {
    "_id": "LF47c5saKFZmw46Dr",
    "open": true,
    "alert": false,
    "unread": 0,
    "userMentions": 0,
    "groupMentions": 0,
    "ts": "2018-04-02T15:03:35.870Z",
    "rid": "Phqoza3yiHbS5jGZm",
    "name": "test-channel",
    "fname": "test-channel",
    "customFields": {},
    "t": "c",
    "u": {
      "_id": "DsjiX7fFCo8sD5moA",
      "username": "samar.sunkaria",
      "name": null
    },
    "ls": "2018-04-02T15:15:15.954Z",
    "_updatedAt": "2018-04-02T15:15:15.954Z",
    "roles": [
      "owner"
    ],
    "meta": {
      "revision": 0,
      "created": 1523707185372,
      "version": 0
    }
  },
  "success": true
}
```

This is what I get **after** changing the counter preference from default (show counter) to **hide counter**.
```
{
  "subscription": {
    "_id": "LF47c5saKFZmw46Dr",
    "open": true,
    "alert": false,
    "unread": 0,
    "userMentions": 0,
    "groupMentions": 0,
    "ts": "2018-04-02T15:03:35.870Z",
    "rid": "Phqoza3yiHbS5jGZm",
    "name": "test-channel",
    "fname": "test-channel",
    "customFields": {},
    "t": "c",
    "u": {
      "_id": "DsjiX7fFCo8sD5moA",
      "username": "samar.sunkaria",
      "name": null
    },
    "ls": "2018-04-02T15:15:15.954Z",
    "_updatedAt": "2018-04-14T22:53:21.153Z",
    "roles": [
      "owner"
    ],
    "meta": {
      "revision": 1,
      "created": 1523707185372,
      "version": 0,
      "updated": 1523746401164
    },
    "hideUnreadStatus": true // New flag
  },
  "success": true
}
```

**Comment by artrmz** [comment]
@Sameesunkaria Oh, thank you for checking it out. It's kinda weird but that's fine ^^

**Comment by artrmz** [comment]
So pretty much what is left to do:
- updating settings (implement drop down menu or eventually use accessory view with picker in there) and implement saveNotifications POST
- localize everything
- decide what to do with getOneSubscription endpoint (I guess we have to implement it anyways to always get fresh settings from server)
- update subscription object after saving settings so Subscriptino object that is assigned to Chat/Messages is always up to date [?]
- write tests

**Comment by artrmz** [comment]
@Sameesunkaria Where will be the entry point for In-app Notifications preferences? If it's in Channel's Info screen then we should already have Subscription object (stored in ChatViewController), just need to decide when to update it. But I am not sure if we should use that object or just always fetch fresh via getOneSubscription? Not sure if we should worry about it not being up to date because someone changed those settings in a web or whatever?

**Comment by Sameesunkaria** [comment]
@artrmz The in-app notification sound preferences should be like any other notification preference. Rn I’m on the fence about if we should just use the desktop sound preferences or add another option to the backend. 

Also, does the channel info store all the notification preferences? 🤔

And yeah these settings are not super critical. So, i believe not fetching for the latest changes would be the way to go, cause it would provide for a better experience. (Given the channel info is updated periodically). But, that’s just my opinion. 🙂

**Comment by artrmz** [comment]
@Sameesunkaria Yup, Subscription object contains all the needed information for the notification preferences. Since the entry point for those preferences in in channel's info screen we already have that object ;) `getOneSubscription` seems to be useful only for updating Subscription object in our app 🤔

**Comment by artrmz** [comment]
@rafaelks @cardoso @Sameesunkaria It's pretty much done. Probably made some stupid things in there so let me know what I could improve in your opinion 🤔 So waiting for reviews.

File: Rocket.Chat/Extensions/EnumExtensions.swift [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+//
+//  Created by Artur Rymarz on 18.04.2018.
+//  Copyright © 2018 Rocket.Chat. All rights reserved.
+//
+
+import Foundation
+
+#if swift(>=4.2)
+fatalError("Remove CaseIterable. This should be implemented in Swift 4.2.")
+#else
```
Comment (rafaelks) [inline Rocket.Chat/Extensions/EnumExtensions.swift]: 🤓

Comment (artrmz) [reply Rocket.Chat/Extensions/EnumExtensions.swift]: 🙈🙊 Could remove that or at least replace with just a warning 🤯

**Comment by artrmz** [comment]
@rafaelks, sure! I think i tried too much to make it Web-like 😀

File: Rocket.Chat/NotificationsSwitchCell.swift [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+        let value: Dynamic<Bool>
+        var type: NotificationCellType
+        let leftTitle: String
+        let leftDescription: String
+        let rightTitle: String
+        let rightDescription: String
+
+        init(value: Dynamic<Bool>, type: NotificationCellType, leftTitle: String, leftDescription: String, rightTitle: String, rightDescription: String) {
+            self.value = value
+            self.type = type
```
Comment (Sameesunkaria) [inline Rocket.Chat/NotificationsSwitchCell.swift]: I believe we don't need this `init`. 🧐

**Comment by Sameesunkaria** [comment]
It would be awesome if we could get some table view animations going! 👀

**Comment by artrmz** [comment]
@rafaelks I have merged develop to it. I might start working on it again if you do not plan to any time soon. I think I will ditch `DropDownMenu` and replace it with UITableViewController (or should it be picker on the bottom?)

**Comment by Sameesunkaria** [comment]
Usually apple does a tableview, in which if you tap a cell, a picker view row is added below. A picker view on the bottom is also a nice option.

hudzk

**Comment by rafaelks** [comment]
@artrmz Sounds great!!! Yes, @Sameesunkaria suggestion is good, I also like this approach 👍 

Let's try to remove all @artrmz dependencies that we can easily do manually, OK? Thank you very much!

**Comment by artrmz** [comment]
Replaced drop-down menu with picker view (updated gif in the first post). Not sure what is wrong with the alert/message view at the end after saving preferences? Did I something wrong or is it bugged and should be fixed in another PR?

**Comment by artrmz** [comment]
Added 2 TODOs to the first post. @rafaelks if we get an icon I can add that, not sure about the second one - does it have to be one in this PR or could be in another one once this one is (if ever :) ) merged?

**Comment by rafaelks** [comment]
@artrmz Yes, let's figure that out... @thiagosanchz will be able to help us. Thiago, we need an icon to display inside the "Chat Information" screen to allow the user to change the notification settings for that specific room... can you please do that? :-)

**Comment by thiagosanchz** [comment]
Hey @artrmz and @rafaelks this is the icon that you need.
[notification_icon.pdf](https://github.com/RocketChat/Rocket.Chat.iOS/files/2207604/notification_icon.pdf)

**Comment by Sameesunkaria** [comment]
Guys this PR is ready for review.

**Comment by rafaelks** [comment]
This is looking really really good, thanks a lot for the work on this!

Few things we can improve before we merge IMHO:

1. The success message is danm big and message is cutting (see attachment);
2. The "Save" button could be enabled only when there are changes; If there are no changes, the button should be disabled;
3. Toggling the "Receive Notifications" is hiding all the cells with no animation at all, can we animate this?

img_5348

**Comment by Sameesunkaria** [comment]
@artrmz Can you please help me with the animations?

**Comment by artrmz** [comment]
@Sameesunkaria Can take a look at that tomorrow morning 👍

**Comment by Sameesunkaria** [comment]
- [x] The success message is danm big and message is cutting (see attachment);
- [ ] The "Save" button could be enabled only when there are changes; If there are no changes, the button should be disabled;
- [ ] Toggling the "Receive Notifications" is hiding all the cells with no animation at all, can we animate this?

File: Rocket.Chat/Resources/cs.lproj/Localizable.strings [hunk 3]
```diff
@@ -366,7 +366,7 @@
 "myaccount.settings.notifications.duration.default" = "Default"; // TODO
 "myaccount.settings.notifications.duration.seconds" = "seconds"; // TODO
 
-"alert.update_notifications_preferences_success.title" = "Successfully updated your notification preferences."; // TODO
+"alert.update_notifications_preferences_success.title" = "Perferences saved"; // TODO
```
Comment (kaiiiiiiiii) [inline Rocket.Chat/Resources/cs.lproj/Localizable.strings]: Typo :) "P**re**ferences saved".

File: Rocket.Chat/Resources/de.lproj/Localizable.strings [hunk 4]
```diff
@@ -366,7 +366,7 @@
 "myaccount.settings.notifications.duration.default" = "Default"; // TODO
 "myaccount.settings.notifications.duration.seconds" = "seconds"; // TODO
 
-"alert.update_notifications_preferences_success.title" = "Successfully updated your notification preferences."; // TODO
+"alert.update_notifications_preferences_success.title" = "Perferences saved"; // TODO
```
Comment (kaiiiiiiiii) [inline Rocket.Chat/Resources/de.lproj/Localizable.strings]: Typo :) "P**re**ferences saved".

File: Rocket.Chat/Resources/el.lproj/Localizable.strings [hunk 5]
```diff
@@ -366,7 +366,7 @@
 "myaccount.settings.notifications.duration.default" = "Default"; // TODO
 "myaccount.settings.notifications.duration.seconds" = "seconds"; // TODO
 
-"alert.update_notifications_preferences_success.title" = "Successfully updated your notification preferences."; // TODO
+"alert.update_notifications_preferences_success.title" = "Perferences saved"; // TODO
```
Comment (kaiiiiiiiii) [inline Rocket.Chat/Resources/el.lproj/Localizable.strings]: Typo :) "P**re**ferences saved".

File: Rocket.Chat/Resources/en.lproj/Localizable.strings [hunk 6]
```diff
@@ -365,7 +365,7 @@
 "myaccount.settings.notifications.duration.default" = "Default";
 "myaccount.settings.notifications.duration.seconds" = "seconds";
 
-"alert.update_notifications_preferences_success.title" = "Successfully updated your notification preferences.";
+"alert.update_notifications_preferences_success.title" = "Perferences saved";
```
Comment (kaiiiiiiiii) [inline Rocket.Chat/Resources/en.lproj/Localizable.strings]: Typo :) "P**re**ferences saved".

File: Rocket.Chat/Resources/es.lproj/Localizable.strings [hunk 7]
```diff
@@ -365,7 +365,7 @@
 "myaccount.settings.notifications.duration.default" = "Default"; // TODO
 "myaccount.settings.notifications.duration.seconds" = "seconds"; // TODO
 
-"alert.update_notifications_preferences_success.title" = "Successfully updated your notification preferences."; // TODO
+"alert.update_notifications_preferences_success.title" = "Perferences saved"; // TODO
```
Comment (kaiiiiiiiii) [inline Rocket.Chat/Resources/es.lproj/Localizable.strings]: Typo :) "P**re**ferences saved".

File: Rocket.Chat/Resources/fr.lproj/Localizable.strings [hunk 8]
```diff
@@ -364,7 +364,7 @@
 "myaccount.settings.notifications.duration.default" = "Default"; // TODO
 "myaccount.settings.notifications.duration.seconds" = "seconds"; // TODO
 
-"alert.update_notifications_preferences_success.title" = "Successfully updated your notification preferences."; // TODO
+"alert.update_notifications_preferences_success.title" = "Perferences saved"; // TODO
```
Comment (kaiiiiiiiii) [inline Rocket.Chat/Resources/fr.lproj/Localizable.strings]: Typo :) "P**re**ferences saved".

File: Rocket.Chat/Resources/pl.lproj/Localizable.strings [hunk 9]
```diff
@@ -365,7 +365,7 @@
 "myaccount.settings.notifications.duration.default" = "Default"; // TODO
 "myaccount.settings.notifications.duration.seconds" = "seconds"; // TODO
 
-"alert.update_notifications_preferences_success.title" = "Successfully updated your notification preferences."; // TODO
+"alert.update_notifications_preferences_success.title" = "Perferences saved"; // TODO
```
Comment (kaiiiiiiiii) [inline Rocket.Chat/Resources/pl.lproj/Localizable.strings]: Typo :) "P**re**ferences saved".

File: Rocket.Chat/Resources/pt-BR.lproj/Localizable.strings [hunk 10]
```diff
@@ -366,7 +366,7 @@
 "myaccount.settings.notifications.duration.default" = "Default"; // TODO
 "myaccount.settings.notifications.duration.seconds" = "seconds"; // TODO
 
-"alert.update_notifications_preferences_success.title" = "Successfully updated your notification preferences."; // TODO
+"alert.update_notifications_preferences_success.title" = "Perferences saved"; // TODO
```
Comment (kaiiiiiiiii) [inline Rocket.Chat/Resources/pt-BR.lproj/Localizable.strings]: Typo :) "P**re**ferences saved".

File: Rocket.Chat/Resources/pt-PT.lproj/Localizable.strings [hunk 11]
```diff
@@ -363,7 +363,7 @@
 "myaccount.settings.notifications.duration.default" = "Default"; // TODO
 "myaccount.settings.notifications.duration.seconds" = "seconds"; // TODO
 
-"alert.update_notifications_preferences_success.title" = "Successfully updated your notification preferences."; // TODO
+"alert.update_notifications_preferences_success.title" = "Perferences saved"; // TODO
```
Comment (kaiiiiiiiii) [inline Rocket.Chat/Resources/pt-PT.lproj/Localizable.strings]: Typo :) "P**re**ferences saved".

File: Rocket.Chat/Resources/ru.lproj/Localizable.strings [hunk 12]
```diff
@@ -365,7 +365,7 @@
 "myaccount.settings.notifications.duration.default" = "Default"; // TODO
 "myaccount.settings.notifications.duration.seconds" = "seconds"; // TODO
 
-"alert.update_notifications_preferences_success.title" = "Successfully updated your notification preferences."; // TODO
+"alert.update_notifications_preferences_success.title" = "Perferences saved"; // TODO
```
Comment (kaiiiiiiiii) [inline Rocket.Chat/Resources/ru.lproj/Localizable.strings]: Typo :) "P**re**ferences saved".

File: Rocket.Chat/Resources/cs.lproj/Localizable.strings (hunk 3 above)
Comment (Sameesunkaria) [reply Rocket.Chat/Resources/cs.lproj/Localizable.strings]: Sometimes I hate that Xcode doesn't have autocorrect for strings. 🙈

File: Rocket.Chat/Resources/el.lproj/Localizable.strings (hunk 5 above)
Comment (cardoso) [reply Rocket.Chat/Resources/el.lproj/Localizable.strings]: still missing this one @Sameesunkaria

**Comment by artrmz** [comment]
@Sameesunkaria I am done with enabling/disabling save button although not sure yet if I want to push that :) We either have to request subscription to make sure we are synced with the server or leave the controller after saving, not sure what is better... additional request or closing controller 🤔 

Will let know once I finish the animation.

**Comment by artrmz** [comment]
@rafaelks @Sameesunkaria Pushed the change to enable/disable "Save button" as needed.

**Comment by rafaelks** [comment]
@artrmz @Sameesunkaria Looking great to me, gonna review the code!

**Comment by Sameesunkaria** [comment]
Thanks @artrmz for all your hard work on this one! 🎉

**Comment by rafaelks** [comment]
Yes, thank you very much for that @artrmz, many people are going to be happy with this PR!!! 🎉

**Pull request merged**